### PR TITLE
Deprecate field type annotation classes in favor of Field

### LINKED ISSUE #1207: Deprecate field type annotation classes in favor of Field
This is related to #1168 (deprecating annotations that share names with reserved words in PHP7). It is more directly related to doctrine/phpcr-odm#641 and the corresponding PR (doctrine/phpcr-odm#643).

We should encourage users to use `@Field(type="...")` instead of the plethora of named annotation classes. Deprecating this in 1.x will allow us to remove the named classes entirely in 2.0.

### PR DESCRIPTION
**Pull request by alcaeus**
Closes #1207.

Starting with PHP 7, some of the type annotation ODM uses are reserved keywords:
* int
* string
* bool
* float

To avoid having large inconsistencies we decided to deprecate all type annotations (e.g. those extending `AbstractField` and only overwriting the `type` property) to be removed in ODM 2.0. This affects the following annotation classes:
* Bin
* BinCustom
* BinFunc
* BinMD5
* BinUUID
* BinUUIDRFC4122
* Bool
* Boolean
* Collection
* Date
* Float
* Hash
* Increment
* Int
* Integer
* Key
* ObjectId
* Raw
* String
* Timestamp

Instead, users should switch to using `@Field(type="...")` to map fields.

### CODE REVIEW

**Comment by malarzm** [comment]
I think [this](https://github.com/doctrine/mongodb-odm/pull/1318/files#diff-1457aa7b8fa8bab45fea926215c23cbaR110) among the others from bigger list should have different notes? 

Beside that big :+1: for going through everything!

**Comment by alcaeus** [comment]
Yes, we could expand the wording, should we also add an example of how the field show be mapped?

**Comment by malarzm** [comment]
> should we also add an example of how the field show be mapped?

Was thinking about it and frankly I'm not sure... existing examples are updated for the new way and we could use that until absolute removal which will need documentation changes anyway so we could postpone writing new chapter? Unless you feel like doing that :D

**Comment by alcaeus** [comment]
I'm not sure we need this. I think for 1.1 just deprecating the classes is a good first step. We can then always throw deprecation notices in `AnnotationDriver` when a deprecated annotation is used.

**Comment by malarzm** [comment]
:+1:

**Comment by malarzm** [comment]
Another :+1: after rewording deprecation notes :)

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Bin.php [hunk 1]
```diff
@@ -19,7 +19,10 @@
 
 namespace Doctrine\ODM\MongoDB\Mapping\Annotations;
 
-/** @Annotation */
+/**
+ * @Annotation
```
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Bin.php]: Is `@Annotation` a convention from Doctrine ORM or Symfony? I'm curious if we should just replace this with a one-line class description (e.g. "Annotation for binary data field.")

**Comment by jmikola** [comment]
One question above, but LGTM otherwise. Thanks for the tedious search/replace job!

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Bin.php (hunk 1 above)
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Bin.php]: IIRC, the `@Annotation` annotation tells the annotation reader that the class in question can be used as an annotation.
I'm actually :+1: to documenting annotation classes and their properties while we're at it.

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Document.php [hunk 2]
```diff
@@ -19,7 +19,11 @@
 
 namespace Doctrine\ODM\MongoDB\Mapping\Annotations;
 
-/** @Annotation */
+/**
+ * Identifies a class as document that can be stored in the database
```
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Document.php]: "as a document"

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/EmbeddedDocument.php [hunk 3]
```diff
@@ -19,7 +19,11 @@
 
 namespace Doctrine\ODM\MongoDB\Mapping\Annotations;
 
-/** @Annotation */
+/**
+ * Identified a class as document that can be embedded but not stored by itself
```
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/EmbeddedDocument.php]: "Identifies"

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/MappedSuperclass.php [hunk 4]
```diff
@@ -19,7 +19,11 @@
 
 namespace Doctrine\ODM\MongoDB\Mapping\Annotations;
 
-/** @Annotation */
+/**
+ * Specify class as parent class for documents containing mapping information
```
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/MappedSuperclass.php]: "Specifieds a parent class that other documents may extend to inherit mapping information"

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/EmbedMany.php [hunk 5]
```diff
@@ -21,7 +21,11 @@
 
 use Doctrine\ODM\MongoDB\Utility\CollectionHelper;
 
-/** @Annotation */
+/**
+ * Specifies a one-to-many relationship to a different document
```
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/EmbedMany.php]: I think this is mixed up with ReferenceMany's doc block.

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/EmbedOne.php [hunk 6]
```diff
@@ -19,7 +19,11 @@
 
 namespace Doctrine\ODM\MongoDB\Mapping\Annotations;
 
-/** @Annotation */
+/**
+ * Specifies a one-to-one relationship to a different document
```
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/EmbedOne.php]: This also appears to be mixed up with ReferenceOne's doc block.

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/ReferenceOne.php [hunk 7]
```diff
@@ -19,7 +19,11 @@
 
 namespace Doctrine\ODM\MongoDB\Mapping\Annotations;
 
-/** @Annotation */
+/**
+ * Embeds a single document
```
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/ReferenceOne.php]: Is this mixed up with EmbedOne's doc block?

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/ReferenceMany.php [hunk 8]
```diff
@@ -21,7 +21,11 @@
 
 use Doctrine\ODM\MongoDB\Utility\CollectionHelper;
 
-/** @Annotation */
+/**
+ * Embeds multiple documents
```
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/ReferenceMany.php]: Is this mixed up with EmbedMany's doc block?

**Pull request closed without merging**